### PR TITLE
fix(go): handle UUID data type in metadata paths

### LINKED ISSUE #164: fix(go): UUID data type not handled in metadata paths (GetTableSchema / GetObjects)
## What happened

Columns of Snowflake's [`UUID` data type](https://docs.snowflake.com/en/sql-reference/data-types-uuid) cannot be discovered through the driver's metadata paths, even though they query fine.

- `GetTableSchema` (which runs `DESC TABLE` → `descToField`) returns a hard error:
  ```
  Not Implemented: Snowflake Data Type UUID not implemented
  ```
- `GetObjects` → `toArrowField` matches no `case` for `UUID`, so the column's `arrow.Field.Type` is left `nil` and no `XdbcDataType`/`XdbcSqlDataType` is populated.

The data path itself is fine: `SELECT * FROM <table>` returns the `UUID` column as Arrow `utf8` (Snowflake delivers UUID values as text over the wire).

This surfaces in the Power BI Snowflake connector (built on this ADBC driver) as a Navigator error when previewing/loading a table that has a `UUID` column:

```
Expression.Error: Unable to understand the type for column
Details:
    <UUID column name>
```

## Reproduction

```sql
create table tester (uuid_id UUID, varchar_id VARCHAR);
insert into tester (uuid_id, varchar_id) select uuid_string(), uuid_string();
```

Then, against `TESTER`:

| Path | Result |
| --- | --- |
| `GetTableSchema` | ❌ `Not Implemented: Snowflake Data Type UUID not implemented` |
| `GetObjects` (column depth) | ⚠️ field type left `nil` |
| `SELECT *` result schema | ✅ `utf8` |

`SHOW COLUMNS` reports the type as `{"type":"UUID","nullable":true}`; `DESC TABLE` reports it as `UUID`.

## Expected

The metadata paths should agree with the data path and report a `UUID` column as Arrow `utf8`, consistent with how `VARIANT`/`OBJECT`/`ARRAY` are surfaced.

## Environment

- `go/` driver, `github.com/snowflakedb/gosnowflake/v2 v2.1.0`

### PR DESCRIPTION
**Pull request by prmoore77**
## Problem

Snowflake's [`UUID` data type](https://docs.snowflake.com/en/sql-reference/data-types-uuid) queries fine but was unhandled in the driver's two schema-discovery paths:

- **`GetTableSchema`** (`DESC TABLE` → `descToField`): a parenless `UUID` type fell into the `default` and returned a hard error — `Not Implemented: Snowflake Data Type UUID not implemented`.
- **`GetObjects`** (`SHOW COLUMNS` → `toArrowField`): no `case` matched `UUID`, so `arrow.Field.Type` was left `nil`.

This surfaced downstream (e.g. the Power BI Snowflake connector, Metabase) as failures like `Expression.Error: Unable to understand the type for column`.

## Fix

UUID columns are surfaced as `arrow.uuid`.

## How UUID result columns are detected

Snowflake does **not** report a dedicated `uuid` type in query-result metadata. Verified against a live account: both the rowtype JSON and the Arrow field metadata report UUID result columns as `text`/`TEXT` (consistent with the docs: "Snowflake drivers treat UUID values as text strings"). Only `DESC TABLE` / `SHOW COLUMNS` name the type — so those paths map `UUID` directly.

On the result path, UUID columns are identified by their otherwise-impossible shape: **`text` with length 0** (VARCHAR's minimum length is 1; SHOW commands, empty strings, `NULL::VARCHAR`, concatenations, etc. all report length ≥ 1, while table columns, `NULL::UUID`, and UUID-typed expressions all report 0). The one other length-0 producer — an untyped `NULL` literal (`SELECT NULL`) — is ruled out on the Arrow data path via its explicit `byteLength: "0"` field metadata (UUID columns carry `"16"` or omit the key). A literal `uuid` rowtype is also matched, to future-proof against the server reporting the logical type directly.

## Changes

- `descToField` (`GetTableSchema`) and `toArrowField` (`GetObjects`) report `arrow.uuid`
- the Arrow data path parses the utf8 values into an `extensions.UUIDArray` via a column transformer; the JSON fallback path types the field so the extension builder parses the text
- **Bulk ingest**: `arrow.uuid` columns are created as native `UUID` columns (previously `BINARY(16)`). On the plain COPY path no transform is needed — Snowflake's Parquet reader converts the Parquet UUID logical type (emitted for arrow.uuid's `fixed_size_binary[16]` storage) natively, verified value-for-value against a live account. On the geo COPY-transform path (where every column is read through a `$1:"col"` VARIANT reference), the 16 raw bytes are re-encoded to canonical UUID text inline.
- **Bind**: `arrow.uuid` parameters (both registered `extensions.UUIDArray` values and C-Data-Interface arrays that arrive as `fixed_size_binary[16]` storage plus `ARROW:extension:name` field metadata) bind as canonical UUID text, which Snowflake converts implicitly.
- Extension detection follows the existing geoarrow pattern: both `arrow.EXTENSION` types and `ARROW:extension:name` field metadata are recognized.

Closes #164

### CODE REVIEW

**Comment by CurtHagenlocher** [comment]
> The Power BI M-layer type mapper handles utf8 but not fixed_size_binary/extension types

That could be addressed.

I feel like the right thing to do for the driver should be independent of a particular consumption scenario. And driver changes should anyway be (ideally) backwards-compatible, e.g. by opting into new behavior with a connection property.

**Comment by prmoore77** [comment]
This is a screenshot of the Power BI connector displaying the error: 
image

**Comment by zeroshade** [comment]
It's frustrating that Snowflake stores the UUID values as fixed size binary values, but still uses the arrow String type for how it returns it :frowning:

**Comment by CurtHagenlocher** [comment]
> This is a screenshot of the Power BI connector displaying the error:

We, um, don't use the ADBC functions to get the table schema. We run `SHOW COLUMNS IN TABLE` and parse the results ourselves, so this is totally a bug in our connector and not the driver.

That's not to say that the driver couldn't be improved in this fashion... .

**Comment by prmoore77** [comment]
> > This is a screenshot of the Power BI connector displaying the error:
> 
> We, um, don't use the ADBC functions to get the table schema. We run `SHOW COLUMNS IN TABLE` and parse the results ourselves, so this is totally a bug in our connector and not the driver.
> 
> That's not to say that the driver couldn't be improved in this fashion... .

Hi, @CurtHagenlocher - Is the connector open-source?  Should I file an issue in that repo (if it is)?

FWIW - We also observed a similar issue in Metabase - and they had to make a fix as well: https://github.com/metabase/metabase/pull/72034

**Comment by CurtHagenlocher** [comment]
> Hi, @CurtHagenlocher - Is the connector open-source? Should I file an issue in that repo (if it is)?

Alas, it is not. I've filed 2175707 in our internal tracking system.

File: go/driver_test.go
Comment (lidavidm) [inline go/driver_test.go:1]: This should be added as a validation suite case

**Review by lidavidm — COMMENTED** [review]
Is this another case where Snowflake doesn't tell us the "real" logical type and so we're unable to attach the proper extension type?

**Comment by zeroshade** [comment]
> Is this another case where Snowflake doesn't tell us the "real" logical type and so we're unable to attach the proper extension type?

It looks like snowflake *does* tell us the "real" logical type in the Column Info and descriptions (but not in the arrow schema) as "UUID", but it still delivers the UUID explicitly as a string and not as a fixed_size_binary[16].

**Comment by lidavidm** [comment]
> > Is this another case where Snowflake doesn't tell us the "real" logical type and so we're unable to attach the proper extension type?
> 
> It looks like snowflake _does_ tell us the "real" logical type in the Column Info and descriptions (but not in the arrow schema) as "UUID", but it still delivers the UUID explicitly as a string and not as a fixed_size_binary[16].

In that case, though, presumably it's not too expensive to cast the data and "fix" it?

**Comment by zeroshade** [comment]
> In that case, though, presumably it's not too expensive to cast the data and "fix" it?

That would be my preference also, but it looks like PowerBI rejects that? Do I have that right @prmoore77 @CurtHagenlocher ? 

Is it possible to change this so that we cast the uuid values to a proper arrow.uuid extension type using fixed size binary[16]?

**Comment by prmoore77** [comment]
> > In that case, though, presumably it's not too expensive to cast the data and "fix" it?
> 
> That would be my preference also, but it looks like PowerBI rejects that? Do I have that right @prmoore77 @CurtHagenlocher ?
> 
> Is it possible to change this so that we cast the uuid values to a proper arrow.uuid extension type using fixed size binary[16]?

Hey @zeroshade - that is fine from my side.  I can close this PR if you would like to make the fix you would recommend - or I could adjust it, but you would know better than I do about how it should be done...

**Comment by zeroshade** [comment]
I'm fine with you adjusting this PR to use the extension type and cast the string we get from Snowflake accordingly

**Comment by prmoore77** [comment]
Hi, @zeroshade @lidavidm - Done — the PR now surfaces UUID columns as the `arrow.uuid` extension type (`fixed_size_binary[16]`) and casts the text values from the wire into it, and the test moved into the validation suite as `queries/type/select/uuid.txtcase`. One important nuance I hit while implementing it:

> It looks like snowflake *does* tell us the "real" logical type in the Column Info and descriptions (but not in the arrow schema) as "UUID"

That turns out to be true only for `DESC TABLE` / `SHOW COLUMNS` (so `GetTableSchema`/`GetObjects` map `UUID` directly). On the **query-result path** there is no `uuid` type anywhere: I verified against a live account that both the rowtype JSON and the Arrow field metadata report UUID result columns as `text`/`TEXT` — consistent with the [UUID docs](https://docs.snowflake.com/en/sql-reference/data-types-uuid): "Snowflake drivers treat UUID values as text strings."

What UUID result columns *do* have is an otherwise-impossible shape: **`text` with length 0**. VARCHAR's minimum length is 1, and I probed a pile of text producers (SHOW commands, empty strings, `NULL::VARCHAR`, `TO_CHAR`, concatenations, `GET_DDL`, …) — all report length ≥ 1, while UUID table columns, `NULL::UUID`, and UUID-typed expressions all report 0. So the data path detects UUID as `text && length == 0` (plus a literal `uuid` rowtype match, to future-proof against the server ever reporting the logical type directly).

The one collision is an untyped `NULL` literal (`SELECT NULL`), which also comes back as `text` length 0. Those are excluded on the Arrow data path via the server's field metadata: untyped NULL carries an explicit `byteLength: "0"`, while UUID columns carry their 16-byte storage width (`"16"`) or omit the key. There's a regression check in `TestUUIDType` asserting `SELECT NULL` stays `utf8`.

I'll grant this is a fingerprint rather than a documented contract — but it's the only signal the server gives us today (similar in spirit to the EWKB sniffing the driver already does for geo columns), and if gosnowflake/the server ever start reporting `uuid` directly, the explicit type match takes over.

The round trip is also covered now: bulk ingest creates native `UUID` columns for `arrow.uuid` data (previously `BINARY(16)`) — no COPY transform needed on the plain path, since Snowflake's Parquet reader converts the Parquet UUID logical type natively — and `arrow.uuid` bind parameters (previously rejected) bind as canonical UUID text. Both come with validation-suite cases (`ingest/uuid`, `type/bind/uuid`) in addition to `type/select/uuid`.

Verified live: the UUID validation cases pass 8/8, the full validation query suite passes 212/212, and the ingest and integration suites show no new failures beyond ones that reproduce identically on `main`.

**Comment by prmoore77** [comment]
Hi @zeroshade and @lidavidm - there is a failing test above, but I think it is due to credentials.

**Comment by zeroshade** [comment]
yea, since this PR is on a fork it can't access the secrets to run it which is why it's failing. I'll run it myself manually to confirm and post back

**Review by zeroshade — APPROVED** [review]
running locally, the validation on this branch passed successfully. I'll wait for @lidavidm to comment before merging, but this looks good to me

**Comment by prmoore77** [comment]
Hi @zeroshade and @lidavidm - is there any chance this fix could get merged in some form?  Having UUID's available in Snowflake via ADBC would be beneficial...  Thanks for your help.

**Comment by prmoore77** [comment]
> > Hi, @CurtHagenlocher - Is the connector open-source? Should I file an issue in that repo (if it is)?
> 
> Alas, it is not. I've filed 2175707 in our internal tracking system.

Hi @CurtHagenlocher - is that issue being worked at all?  (just curious - asking for a friend 😊)

File: go/validation/queries/type/select/uuid.txtcase [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+;
+
+// part: query
+
+SELECT RES FROM TEST_UUID ORDER BY IDX ASC
+
+// part: expected_schema
+// Snowflake delivers UUID values as utf8 text on the wire; the driver decodes
+// them into the arrow.uuid extension type (fixed_size_binary[16] storage) and
+// reports the same type from the metadata paths.
```
Comment (lidavidm) [inline go/validation/queries/type/select/uuid.txtcase]:
```suggestion
```

Can we clean up some of the AI comments?

File: go/driver_test.go
Comment (lidavidm) [reply go/driver_test.go:1]: ...and now that the cases are added, is this still necessary?

File: go/record_reader.go [hunk 2]
```diff
@@ -165,6 +166,53 @@ func stripEWKBSRIDColumn(_ context.Context, a arrow.Array) (arrow.Array, error)
 	return bldr.NewArray(), nil
 }
 
+// isUUIDResultColumn reports whether a result column carries Snowflake UUID
+// values. Snowflake result metadata does not surface a dedicated "uuid" type
+// today: UUID columns are reported as "text" but with length 0, which no real
+// text column can have (VARCHAR's minimum length is 1, and even SHOW commands
+// report the maximum length for their text columns). The explicit "uuid"
+// match future-proofs against the server reporting the logical type directly.
```
Comment (lidavidm) [inline go/record_reader.go]: If this is never actually used, I don't think it makes sense to check for it. There's no guarantee the implementation would look like this anyways.

File: go/record_reader.go [hunk 3]
```diff
@@ -600,6 +657,13 @@ func rowTypesToArrowSchema(_ context.Context, ld gosnowflake.ArrowStreamLoader,
 				[]string{srcMeta.Type},
 			),
 		}
+		if isUUIDResultColumn(srcMeta.Type, srcMeta.Length) {
+			// jsonDataToArrow appends the JSON UUID text through the
+			// extension builder's AppendValueFromString, which parses it
+			// into the fixed_size_binary[16] storage.
```
Comment (lidavidm) [inline go/record_reader.go]:
```suggestion
```

File: go/record_reader.go [hunk 4]
```diff
@@ -298,6 +346,15 @@ func getTransformer(sc *arrow.Schema, ld gosnowflake.ArrowStreamLoader, useHighP
 			continue
 		}
 
+		// Snowflake delivers UUID values as utf8 text on the wire; decode
+		// them into the canonical arrow.uuid extension type.
```
Comment (lidavidm) [inline go/record_reader.go]:
```suggestion
```

**Comment by lidavidm** [comment]
(Also, if we use the [workflow](https://github.com/adbc-drivers/snowflake/actions/workflows/go_test_pr.yaml), then the test results will be posted to the PR.)

**Comment by CurtHagenlocher** [comment]
> Hi @CurtHagenlocher - is that issue being worked at all? (just curious - asking for a friend 😊)

I'm going to try to get to it this week.

File: go/record_reader.go (hunk 2 above)
Comment (prmoore77) [reply go/record_reader.go]: Removed the speculative uuid type-name match; the function now only checks the shape Snowflake actually reports today

**Comment by prmoore77** [comment]
@lidavidm - for `driver_test.go`: Removed both integration tests as redundant with the validation cases; kept only a small TestUntypedNullLiteral regression test for the SELECT NULL edge case, which the validation  
  suite can't express.
  
  (I wasn't sure how to reply in-thread for this)

**Review by lidavidm — APPROVED** [review]

**Comment by prmoore77** [comment]
🚀
